### PR TITLE
fix: migrated to 2018 edition

### LINKED ISSUE #300: Migrate to 2018 edition
We should probably migrate a bit before it is released on stable.

### PR DESCRIPTION
**Pull request by tomtau**
closes #300

the code was migrated via `cargo fix --edition`
and the grammar benchmarks were rewritten to use `bencher`
instead of the nightly test (so that they can be built
on a stable version).

### CODE REVIEW

**Comment by glyn** [comment]
What is your rationale for migrating to the 2018 edition rather than 2021?

**Comment by tomtau** [comment]
@glyn I was originally planning 2021, but it seems the current minimal version on CI doesn't support it: https://github.com/pest-parser/pest/runs/7268848504?check_suite_focus=true hence I switched it to 2018.

maybe good for @CAD97 to comment whether it's ok to bump up MSRV.

**Comment by birkenfeld** [comment]
IMO: don't bump MSRV for such a minor thing, the 2021 edition probably doesn't have anything to offer `pest`.

**Comment by CAD97** [comment]
If we bump MSRV it should be in a separate patch. MSRV should ideally only be increased for a specific benefit; I think it's reasonable to set an MSRV of 1.56 for the [`rust-version` manifest key](https://doc.rust-lang.org/cargo/reference/manifest.html#the-rust-version-field). 1.56 will be a year old on October 21; stable-6 is a common community definition of LTS.

For reference:

distro | version | release date
-- | --
[debian oldstable](https://packages.debian.org/buster/rustc) | 1.41 | Jan 30 2020
[debian stable] | 1.48 | Nov 19 2020
[debian nextstable](https://packages.debian.org/testing/rust/rustc) | 1.59 | Feb 24 2022
[debian unstable](https://packages.debian.org/unstable/rust/rustc) | 1.59 | Feb 24 2022
tokio (tested) | 1.49 | Dec 31 2020
tokio (guaranteed) | stable-6 (1.56) | Oct 21 2021

File: pest/src/span.rs [hunk 1]
```diff
@@ -265,6 +265,16 @@ mod tests {
     use alloc::borrow::ToOwned;
     use alloc::vec::Vec;
 
+    #[test]
```
Comment (CAD97) [inline pest/src/span.rs:268]: These changes seem to have snuck in? Not an issue though, I'm not strict about a perfectly clean git history.

**Review by CAD97 — APPROVED** [review]
Personally I strongly prefer using criterion, but the choice to use bencher here for a smaller diff is reasonable.

Diff is the mechanical edition update, as described.

File: pest/src/span.rs (hunk 1 above)
Comment (tomtau) [reply pest/src/span.rs:268]: Oh yes, one extra test got in

**Comment by tomtau** [comment]
Thanks @CAD97 . I originally used criterion, but it failed on CI due to one of its dependencies requiring edition 2021, hence bencher.
I guess if MSRV is bumped to stable-6, it should be ok to switch to criterion.

**Comment by CAD97** [comment]
It should be fine to only run benchmarks on current stable rather than MSRV, so it shouldn't fail as a dev-dependency.

**Pull request merged**